homeassistant/components/media_player/firetv.py

```diff
@@ -4,166 +4,147 @@
-                    self.proto, self.host, self.port, self.device_id
-                    ), timeout=10).json()
-            return response.get('state', STATE_UNKNOWN)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve device state for %s", self.device_id)
-            return STATE_UNKNOWN
-
-    @property
-    def current_app(self):
-        """Return the current app."""
-        try:
-            response = requests.get(
-                DEVICE_APPS_URL.format(
-                    self.proto, self.host, self.port, self.device_id, 'current'
-                    ), timeout=10).json()
-            _current_app = response.get('current_app')
-            if _current_app:
-                return _current_app.get('package')
+class FireTVDevice(MediaPlayerDevice):
+    """Representation of an Amazon Fire TV device on the network."""
 
-            return None
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve current app for %s", self.device_id)
-            return None
+    def __init__(self, ftv, name, get_source, get_sources):
+        """Initialize the FireTV device."""
+        from adb.adb_protocol import (
+            InvalidCommandError, InvalidResponseError, InvalidChecksumError)
 
-    @property
-    def running_apps(self):
-        """Return a list of running apps."""
-        try:
-            response = requests.get(
-                DEVICE_APPS_URL.format(
-                    self.proto, self.host, self.port, self.device_id, 'running'
-                    ), timeout=10).json()
-            return response.get('running_apps')
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve running apps for %s", self.device_id)
-            return None
+        self.firetv = ftv
 
-    def action(self, action_id):
-        """Perform an action on the device."""
-        try:
-            requests.get(DEVICE_ACTION_URL.format(
-                self.proto, self.host, self.port, self.device_id, action_id
-                ), timeout=10)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Action request for %s was not accepted for device %s",
-                action_id, self.device_id)
-
-    def start_app(self, app_name):
-        """Start an app."""
-        try:
-            requests.get(DEVICE_APPS_URL.format(
-                self.proto, self.host, self.port, self.device_id,
-                app_name + '/start'), timeout=10)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not start %s on %s", app_name, self.device_id)
+        self._name = name
+        self._get_source = get_source
+        self._get_sources = get_sources
 
+        # whether or not the ADB connection is currently in use
+        self.adb_lock = False
```

Comment:
Make this a non blocking `threading.Lock` instead, and acquire and release it in the decorator.

Reply:
Fixed in https://github.com/home-assistant/home-assistant/pull/17767/commits/e14e686b6937c969773fa580bd8b7200664e90d4